// @wormhole-foundation/sdk-definitions/dist/esm/protocols/executor/api.js
import { toChainId } from "@wormhole-foundation/sdk-base";
export var RelayStatus;
(function (RelayStatus) {
    RelayStatus["Pending"] = "pending";
    RelayStatus["Failed"] = "failed";
    RelayStatus["Unsupported"] = "unsupported";
    RelayStatus["Submitted"] = "submitted";
    RelayStatus["Underpaid"] = "underpaid";
    RelayStatus["Aborted"] = "aborted";
})(RelayStatus || (RelayStatus = {}));
export var RequestPrefix;
(function (RequestPrefix) {
    RequestPrefix["ERM1"] = "ERM1";
    RequestPrefix["ERV1"] = "ERV1";
    RequestPrefix["ERN1"] = "ERN1";
    RequestPrefix["ERC1"] = "ERC1";
    RequestPrefix["ERC2"] = "ERC2";
})(RequestPrefix || (RequestPrefix = {}));
export async function fetchCapabilities(apiBaseUrl) {
    const url = `${apiBaseUrl}/capabilities`;
    try {
        const response = await fetch(url, {
            method: "GET",
            headers: {
                "Content-Type": "application/json",
            },
        });
        if (!response.ok) {
            throw new Error(`HTTP error! status: ${response.status}`);
        }
        return (await response.json());
    }
    catch (error) {
        throw new Error("Failed to fetch capabilities.");
    }
}
export async function fetchQuote(apiBaseUrl, srcChain, dstChain, relayInstructions) {
    const url = `${apiBaseUrl}/quote`;
    try {
        const response = await fetch(url, {
            method: "POST",
            headers: {
                "Content-Type": "application/json",
            },
            body: JSON.stringify({
                srcChain: toChainId(srcChain),
                dstChain: toChainId(dstChain),
                relayInstructions,
            }),
        });
        if (!response.ok) {
            throw new Error(`HTTP error! status: ${response.status}`);
        }
        return (await response.json());
    }
    catch (error) {
        throw new Error(`Failed to fetch signed quote.`);
    }
}
export async function fetchStatus(apiBaseUrl, txHash, chain) {
    const url = `${apiBaseUrl}/status/tx`;
    try {
        const response = await fetch(url, {
            method: "POST",
            headers: {
                "Content-Type": "application/json",
            },
            body: JSON.stringify({
                txHash,
                chainId: toChainId(chain),
            }),
        });
        if (!response.ok) {
            throw new Error(`HTTP error! status: ${response.status}`);
        }
        return (await response.json());
    }
    catch (error) {
        throw new Error(`Failed to fetch status for txHash: ${txHash}.`);
    }
}
