// @wormhole-foundation/sdk-definitions/dist/cjs/protocols/executor/api.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.fetchStatus = exports.fetchQuote = exports.fetchCapabilities = exports.RequestPrefix = exports.RelayStatus = void 0;
const sdk_base_1 = require("@wormhole-foundation/sdk-base");
var RelayStatus;
(function (RelayStatus) {
    RelayStatus["Pending"] = "pending";
    RelayStatus["Failed"] = "failed";
    RelayStatus["Unsupported"] = "unsupported";
    RelayStatus["Submitted"] = "submitted";
    RelayStatus["Underpaid"] = "underpaid";
    RelayStatus["Aborted"] = "aborted";
})(RelayStatus || (exports.RelayStatus = RelayStatus = {}));
var RequestPrefix;
(function (RequestPrefix) {
    RequestPrefix["ERM1"] = "ERM1";
    RequestPrefix["ERV1"] = "ERV1";
    RequestPrefix["ERN1"] = "ERN1";
    RequestPrefix["ERC1"] = "ERC1";
    RequestPrefix["ERC2"] = "ERC2";
})(RequestPrefix || (exports.RequestPrefix = RequestPrefix = {}));
async function fetchCapabilities(apiBaseUrl) {
    const url = `${apiBaseUrl}/capabilities`;
    try {
        const response = await fetch(url, {
            method: "GET",
            headers: {
                "Content-Type": "application/json",
            },
        });
        if (!response.ok) {
            throw new Error(`HTTP error! status: ${response.status}`);
        }
        return (await response.json());
    }
    catch (error) {
        throw new Error("Failed to fetch capabilities.");
    }
}
exports.fetchCapabilities = fetchCapabilities;
async function fetchQuote(apiBaseUrl, srcChain, dstChain, relayInstructions) {
    const url = `${apiBaseUrl}/quote`;
    try {
        const response = await fetch(url, {
            method: "POST",
            headers: {
                "Content-Type": "application/json",
            },
            body: JSON.stringify({
                srcChain: (0, sdk_base_1.toChainId)(srcChain),
                dstChain: (0, sdk_base_1.toChainId)(dstChain),
                relayInstructions,
            }),
        });
        if (!response.ok) {
            throw new Error(`HTTP error! status: ${response.status}`);
        }
        return (await response.json());
    }
    catch (error) {
        throw new Error(`Failed to fetch signed quote.`);
    }
}
exports.fetchQuote = fetchQuote;
async function fetchStatus(apiBaseUrl, txHash, chain) {
    const url = `${apiBaseUrl}/status/tx`;
    try {
        const response = await fetch(url, {
            method: "POST",
            headers: {
                "Content-Type": "application/json",
            },
            body: JSON.stringify({
                txHash,
                chainId: (0, sdk_base_1.toChainId)(chain),
            }),
        });
        if (!response.ok) {
            throw new Error(`HTTP error! status: ${response.status}`);
        }
        return (await response.json());
    }
    catch (error) {
        throw new Error(`Failed to fetch status for txHash: ${txHash}.`);
    }
}
exports.fetchStatus = fetchStatus;
