// @wormhole-foundation/sdk-definitions/dist/cjs/protocols/executor/relayInstruction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.relayInstructionsLayout = exports.relayInstructionLayout = exports.gasDropOffInstructionLayout = exports.gasInstructionLayout = void 0;
const universalAddress_js_1 = require("../../layout-items/universalAddress.js");
exports.gasInstructionLayout = [
    { name: "gasLimit", binary: "uint", size: 16 },
    { name: "msgValue", binary: "uint", size: 16 },
];
exports.gasDropOffInstructionLayout = [
    { name: "dropOff", binary: "uint", size: 16 },
    { name: "recipient", ...universalAddress_js_1.universalAddressItem },
];
exports.relayInstructionLayout = [
    {
        name: "request",
        binary: "switch",
        idSize: 1,
        idTag: "type",
        layouts: [
            [[1, "GasInstruction"], exports.gasInstructionLayout],
            [[2, "GasDropOffInstruction"], exports.gasDropOffInstructionLayout],
        ],
    },
];
exports.relayInstructionsLayout = [
    {
        name: "requests",
        binary: "array",
        layout: exports.relayInstructionLayout,
    },
];
