// @wormhole-foundation/sdk-definitions/dist/cjs/protocols/executor/signedQuote.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.signedQuoteLayout = exports.quoteLayout = void 0;
const date_js_1 = require("../../layout-items/date.js");
exports.quoteLayout = [
    {
        name: "quote",
        binary: "switch",
        idSize: 4,
        idTag: "prefix",
        layouts: [
            [
                [0x45513031, "EQ01"],
                [
                    {
                        name: "quoterAddress",
                        binary: "bytes",
                        size: 20,
                    },
                    {
                        name: "payeeAddress",
                        binary: "bytes",
                        size: 32,
                    },
                    { name: "srcChain", binary: "uint", size: 2 },
                    { name: "dstChain", binary: "uint", size: 2 },
                    {
                        name: "expiryTime",
                        binary: "uint",
                        size: 8,
                        custom: date_js_1.dateConversion,
                    },
                    { name: "baseFee", binary: "uint", size: 8 },
                    { name: "dstGasPrice", binary: "uint", size: 8 },
                    { name: "srcPrice", binary: "uint", size: 8 },
                    { name: "dstPrice", binary: "uint", size: 8 },
                ],
            ],
        ],
    },
];
exports.signedQuoteLayout = [
    ...exports.quoteLayout,
    { name: "signature", binary: "bytes", size: 65 },
];
