// hfv-sdk/dist/index.mjs
// src/config.ts
var HFV_API_BASE_URL = process.env.HFV_API_BASE_URL || "https://hfv-api.onrender.com";
var defaultConfig = {
  env: "mainnet",
  apiBaseUrl: "https://api.hfvprotocol.org"
};

// src/client.ts
var HFVClient = class {
  constructor(userConfig = {}) {
    this.config = {
      ...defaultConfig,
      ...userConfig
    };
  }
  get apiBaseUrl() {
    return this.config.apiBaseUrl;
  }
  get env() {
    return this.config.env;
  }
};

// src/bridge/bridgeClient.ts
import axios from "axios";
var HFVBridge = class {
  constructor(client) {
    this.client = client;
    this.http = axios.create({
      baseURL: client.apiBaseUrl || "https://hfv-api.onrender.com",
      timeout: 2e4
    });
  }
  /**
   * Request a bridge quote.
   * POST /bridge/quote
   */
  async getQuote(params) {
    const payload = { ...params, env: this.client.env };
    const { data } = await this.http.post("/bridge/quote", payload);
    return {
      estimatedGasUsd: Number(data.estimatedGasUsd ?? 0),
      estimatedOutputAmount: String(data.estimatedOutputAmount ?? "0"),
      routeName: data.routeName ?? "default",
      feeBreakdown: data.feeBreakdown ?? {},
      quoteId: data.quoteId
    };
  }
  /**
   * Execute the bridge.
   * POST /bridge/execute
   */
  async bridge(params) {
    const payload = { ...params, env: this.client.env };
    const { data } = await this.http.post("/bridge/execute", payload);
    return {
      trackingId: data.trackingId,
      sourceTxHash: data.sourceTxHash
    };
  }
  /**
   * Track the bridge status.
   * GET /bridge/status?id=...
   */
  async getStatus(trackingId) {
    const { data } = await this.http.get("/bridge/status", {
      params: { id: trackingId, env: this.client.env }
    });
    return {
      status: data.status,
      fromChain: data.fromChain,
      toChain: data.toChain,
      sourceTxHash: data.sourceTxHash,
      destTxHash: data.destTxHash,
      errorMessage: data.errorMessage
    };
  }
};

// src/prices/pricesClient.ts
import axios2 from "axios";

// src/chains/chainConfig.ts
var SUPPORTED_CHAINS = [
  {
    chainId: 1,
    key: "ethereum",
    name: "Ethereum",
    symbol: "ETH",
    nativeCoingeckoId: "ethereum",
    logo: "/logos/chains/ethereum.png",
    supportedByWormhole: true,
    wormholeChain: "Ethereum"
  },
  {
    chainId: 10,
    key: "optimism",
    name: "Optimism",
    symbol: "ETH",
    nativeCoingeckoId: "ethereum",
    logo: "/logos/chains/optimism.png",
    supportedByWormhole: true,
    wormholeChain: "Optimism"
  },
  {
    chainId: 56,
    key: "bsc",
    name: "BNB Chain",
    symbol: "BNB",
    nativeCoingeckoId: "binancecoin",
    logo: "/logos/chains/bsc.png",
    supportedByWormhole: true,
    wormholeChain: "BNB Chain"
  },
  {
    chainId: 100,
    key: "gnosis",
    name: "Gnosis",
    symbol: "xDAI",
    nativeCoingeckoId: "xdai",
    logo: "/logos/chains/gnosis.png",
    supportedByWormhole: false
  },
  {
    chainId: 137,
    key: "polygon",
    name: "Polygon",
    symbol: "MATIC",
    nativeCoingeckoId: "matic-network",
    logo: "/logos/chains/polygon.png",
    supportedByWormhole: true,
    wormholeChain: "Polygon"
  },
  {
    chainId: 250,
    key: "fantom",
    name: "Fantom",
    symbol: "FTM",
    nativeCoingeckoId: "fantom",
    logo: "/logos/chains/fantom.png",
    supportedByWormhole: true,
    wormholeChain: "Fantom"
  },
  {
    chainId: 8453,
    key: "base",
    name: "Base",
    symbol: "ETH",
    nativeCoingeckoId: "ethereum",
    logo: "/logos/chains/base.png",
    supportedByWormhole: true,
    wormholeChain: "Base"
  },
  {
    chainId: 59144,
    key: "linea",
    name: "Linea",
    symbol: "ETH",
    nativeCoingeckoId: "ethereum",
    logo: "/logos/chains/linea.png",
    supportedByWormhole: false
  },
  {
    chainId: 34443,
    key: "mode",
    name: "Mode",
    symbol: "ETH",
    nativeCoingeckoId: "ethereum",
    logo: "/logos/chains/mode.png",
    supportedByWormhole: false
  },
  {
    chainId: 42161,
    key: "arbitrum",
    name: "Arbitrum One",
    symbol: "ETH",
    nativeCoingeckoId: "ethereum",
    logo: "/logos/chains/arbitrum.png",
    supportedByWormhole: true,
    wormholeChain: "Arbitrum"
  },
  {
    chainId: 43114,
    key: "avalanche",
    name: "Avalanche C-Chain",
    symbol: "AVAX",
    nativeCoingeckoId: "avalanche-2",
    logo: "/logos/chains/avalanche.png",
    supportedByWormhole: true,
    wormholeChain: "Avalanche"
  },
  {
    chainId: 1329,
    key: "sei",
    name: "Sei",
    symbol: "SEI",
    nativeCoingeckoId: "sei-network",
    logo: "/logos/chains/sei.png",
    supportedByWormhole: true,
    wormholeChain: "Sei"
  },
  {
    chainId: 1313161554,
    key: "aurora",
    name: "Aurora",
    symbol: "ETH",
    nativeCoingeckoId: "ethereum",
    logo: "/logos/chains/aurora.png",
    supportedByWormhole: true,
    wormholeChain: "Aurora"
  },
  {
    chainId: 42220,
    key: "celo",
    name: "Celo",
    symbol: "CELO",
    nativeCoingeckoId: "celo",
    logo: "/logos/chains/celo.png",
    supportedByWormhole: true,
    wormholeChain: "Celo"
  },
  {
    chainId: 1284,
    key: "moonbeam",
    name: "Moonbeam",
    symbol: "GLMR",
    nativeCoingeckoId: "moonbeam",
    logo: "/logos/chains/moonbeam.png",
    supportedByWormhole: true,
    wormholeChain: "Moonbeam"
  },
  {
    chainId: 1285,
    key: "moonriver",
    name: "Moonriver",
    symbol: "MOVR",
    nativeCoingeckoId: "moonriver",
    logo: "/logos/chains/moonriver.png",
    supportedByWormhole: false
  },
  {
    chainId: 16666e5,
    key: "harmony",
    name: "Harmony",
    symbol: "ONE",
    nativeCoingeckoId: "harmony",
    logo: "/logos/chains/harmony.png",
    supportedByWormhole: false
  },
  {
    chainId: 170,
    key: "unichain",
    name: "Unichain",
    symbol: "ETH",
    nativeCoingeckoId: "ethereum",
    logo: "/logos/chains/unichain.png",
    supportedByWormhole: false
  },
  {
    chainId: 7777777,
    key: "zora",
    name: "Zora",
    symbol: "ETH",
    nativeCoingeckoId: "ethereum",
    logo: "/logos/chains/zora.png",
    supportedByWormhole: false
  },
  {
    chainId: 5e3,
    key: "mantle",
    name: "Mantle",
    symbol: "MNT",
    nativeCoingeckoId: "mantle",
    logo: "/logos/chains/mantle.png",
    supportedByWormhole: true,
    wormholeChain: "Mantle"
  },
  {
    chainId: 14,
    key: "flare",
    name: "Flare",
    symbol: "FLR",
    nativeCoingeckoId: "flare-networks",
    logo: "/logos/chains/flare.png",
    supportedByWormhole: false
  },
  {
    chainId: 40,
    key: "telos",
    name: "Telos",
    symbol: "TLOS",
    nativeCoingeckoId: "telos",
    logo: "/logos/chains/telos.png",
    supportedByWormhole: false
  },
  {
    chainId: 50,
    key: "xdc",
    name: "XDC Network",
    symbol: "XDC",
    nativeCoingeckoId: "xdce-crowd-sale",
    logo: "/logos/chains/xdc.png",
    supportedByWormhole: false
  },
  {
    chainId: 57,
    key: "syscoin",
    name: "Syscoin",
    symbol: "SYS",
    nativeCoingeckoId: "syscoin",
    logo: "/logos/chains/syscoin.png",
    supportedByWormhole: false
  },
  {
    chainId: 61,
    key: "etc",
    name: "Ethereum Classic",
    symbol: "ETC",
    nativeCoingeckoId: "ethereum-classic",
    logo: "/logos/chains/etc.png",
    supportedByWormhole: false
  },
  {
    chainId: 57073,
    key: "ink",
    name: "Inkonchain",
    symbol: "INK",
    nativeCoingeckoId: "inkonchain",
    logo: "/logos/chains/ink.png",
    supportedByWormhole: false
  },
  {
    chainId: 122,
    key: "fuse",
    name: "Fuse",
    symbol: "FUSE",
    nativeCoingeckoId: "fuse",
    logo: "/logos/chains/fuse.png",
    supportedByWormhole: false
  },
  {
    chainId: 60808,
    key: "bob",
    name: "BOB",
    symbol: "BOB",
    nativeCoingeckoId: "bob",
    logo: "/logos/chains/bob.png",
    supportedByWormhole: false
  },
  {
    chainId: 81457,
    key: "blast",
    name: "Blast",
    symbol: "ETH",
    nativeCoingeckoId: "blast",
    logo: "/logos/chains/blast.png",
    supportedByWormhole: true,
    wormholeChain: "Blast"
  },
  {
    chainId: 1868,
    key: "soneium",
    name: "Soneium",
    symbol: "ETH",
    nativeCoingeckoId: "soneium",
    logo: "/logos/chains/soneium.png",
    supportedByWormhole: false
  },
  {
    chainId: 480,
    key: "worldchain",
    name: "World Chain",
    symbol: "WLD",
    nativeCoingeckoId: "worldcoin",
    logo: "/logos/chains/worldchain.png",
    supportedByWormhole: true,
    wormholeChain: "World Chain"
  },
  {
    chainId: 1135,
    key: "lisk",
    name: "Lisk",
    symbol: "LSK",
    nativeCoingeckoId: "lisk",
    logo: "/logos/chains/lisk.png",
    supportedByWormhole: false
  },
  {
    chainId: 1923,
    key: "swellchain",
    name: "Swellchain",
    symbol: "SWELL",
    nativeCoingeckoId: "swellchain",
    logo: "/logos/chains/swellchain.png",
    supportedByWormhole: false
  },
  {
    chainId: 2741,
    key: "abstract",
    name: "Abstract",
    symbol: "ABST",
    nativeCoingeckoId: "abstract",
    logo: "/logos/chains/abstract.png",
    supportedByWormhole: false
  },
  {
    chainId: 747474,
    key: "katana",
    name: "Katana",
    symbol: "KATA",
    nativeCoingeckoId: "katana",
    logo: "/logos/chains/katana.png",
    supportedByWormhole: false
  },
  {
    chainId: 146,
    key: "sonic",
    name: "Sonic",
    symbol: "S",
    nativeCoingeckoId: "sonic",
    logo: "/logos/chains/sonic.png",
    supportedByWormhole: false
  }
];

// src/chains/chainUtils.ts
function getChainById(chainId) {
  return SUPPORTED_CHAINS.find((c) => c.chainId === chainId);
}
function getChainByKey(key) {
  return SUPPORTED_CHAINS.find((c) => c.key === key);
}
function getAllSupportedChains() {
  return SUPPORTED_CHAINS;
}
function getWormholeSupportedChains() {
  return SUPPORTED_CHAINS.filter((c) => c.supportedByWormhole);
}

// src/prices/pricesClient.ts
var HFVPrices = class {
  constructor(client) {
    this.client = client;
    this.http = axios2.create({
      baseURL: client.apiBaseUrl || "https://hfv-api.onrender.com",
      timeout: 15e3
    });
  }
  /**
   * Get USD prices for native tokens across chains.
   * GET /prices/native?chainIds=1,56,137
   */
  async getNativePrices(chainIds) {
    const ids = chainIds?.length ? chainIds : getAllSupportedChains().map((c) => c.chainId);
    if (!ids.length) return {};
    const param = ids.join(",");
    const { data } = await this.http.get("/prices/native", {
      params: { chainIds: param }
    });
    return data?.prices || {};
  }
  /**
   * Get USD prices for specific ERC-20 tokens.
   * POST /prices/tokens
   */
  async getTokenPrices(chainId, addresses) {
    if (!chainId || !addresses.length) return {};
    const { data } = await this.http.post("/prices/tokens", {
      chainId,
      addresses
    });
    return data?.prices || {};
  }
};

// src/tokenRegistry.ts
import fs from "fs";
import path from "path";
var TOKENS_DIR = path.join(__dirname, "tokens");
function loadTokens() {
  const registry = {};
  const files = fs.readdirSync(TOKENS_DIR).filter((f) => f.endsWith(".json"));
  for (const file of files) {
    const chainName = file.replace(".json", "");
    const fullPath = path.join(TOKENS_DIR, file);
    try {
      const content = JSON.parse(fs.readFileSync(fullPath, "utf8"));
      if (!Array.isArray(content)) {
        registry[chainName] = [];
        continue;
      }
      registry[chainName] = content;
    } catch (err) {
      console.error(`\u274C Failed to parse token file: ${file}`, err);
      registry[chainName] = [];
    }
  }
  return registry;
}
var tokenRegistry = loadTokens();
export {
  HFVBridge,
  HFVClient,
  HFVPrices,
  HFV_API_BASE_URL,
  SUPPORTED_CHAINS,
  defaultConfig,
  getAllSupportedChains,
  getChainById,
  getChainByKey,
  getWormholeSupportedChains,
  tokenRegistry
};
